### PR TITLE
Rpcprovider

### PR DESCRIPTION
**Pull request by MostafaBalata**
Created new component `RPCProvider` which contains the web3 connection, and if there's no connection will show you message there's no connection. the message component still need some ui

### CODE REVIEW

File: src/containers/Scan.js [hunk 1]
```diff
@@ -39,7 +39,8 @@ class Scan extends Component {
                     <Grid className="scanbox ico-box-scan">
                         <ICO ico={this.props.ico} inner={true} address={this.props.address}/>
                         <ScanBoxLoadingMessage show={this.props.showLoaderState}/>
-                        <ScanBoxDetails address={this.props.address}/>
+
+                        {this.props.currencyValue && <ScanBoxDetails address={this.props.address}/>}
```
Comment (rudolfix) [inline src/containers/Scan.js]: please change to a boolean flag that says that data is loaded

File: src/components/ICO.js [hunk 2]
```diff
@@ -1,19 +1,30 @@
...
+
+  componentDidMount() {
+    this.props.readSmartContract(this.props.address)
+  }
+
+  render() {
+    return (
+      <div>
+        {this.props.inner && <ICOScan address={this.props.address} onModalShow={this.props.onModalShow}/>}
+        {!this.props.inner && <ICOApp address={this.props.address} onModalShow={this.props.onModalShow}/>}
```
Comment (rudolfix) [inline src/components/ICO.js]: rename isInSingleICOView

File: src/actions/CurrencyAction.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import { getEtherPerCurrency, toPromise } from '../utils';
 
-export const setCurrency = (currency, time = new Date().yyyymmdd(), dispatch) => {
+export const setCurrency = (currency, time = new Date().yyyymmdd(), dispatch, callback) => {
   let currencyFormat = `ETH-${currency}`;
   if (time === 'NOW') { time = new Date().yyyymmdd(); }
```
Comment (rudolfix) [inline src/actions/CurrencyAction.js]: why are you using 'NOW'? can't you just leave it null so time = new Date().yyyymmdd() will be correctly assigned?

File: src/actions/CurrencyAction.js [hunk 4]
```diff
@@ -11,50 +11,15 @@ export const setCurrency = (currency, time = new Date().yyyymmdd(), dispatch) =>
       value: 1,
       time,
     });
+
+    // TODO: needs to be correct value
   } else if (currency === 'BTC') { currencyFormat = 'BTC-EUR'; }
```
Comment (rudolfix) [inline src/actions/CurrencyAction.js]: there is no ETH-BTC pair?

File: src/actions/CurrencyAction.js [hunk 5]
```diff
@@ -11,50 +11,15 @@ export const setCurrency = (currency, time = new Date().yyyymmdd(), dispatch) =>
...
 
   const result = getEtherPerCurrency(currencyFormat, time);
   result.then((e) => {
     dispatch({ type: 'SET_CURRENCY', currency, value: e.data.data.amount, time });
+    if (callback) callback();
   }).catch((error) => {
```
Comment (rudolfix) [inline src/actions/CurrencyAction.js]: in web3 reducer you never handle error callback!

File: src/components/ICO.js [hunk 6]
```diff
@@ -7,13 +7,24 @@ import { readSmartContract } from '../reducers/web3';
...
+  }
+
+  componentDidMount() {
+    this.props.readSmartContract(this.props.address);
+  }
+
+  render() {
+    return (
+      <div>
+        {this.props.inner && <ICOScan address={this.props.address} onModalShow={this.props.onModalShow} />}
```
Comment (rudolfix) [inline src/components/ICO.js]: rename inner as specified in other comment

File: src/components/RPCProvider.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import TopHeader from '../components/TopHeader';
+import { default as config } from '../config.js';
+
+const RPCConnectionError = ({ ...props }) => (
+  <div>
+    <TopHeader />
+    <Grid fluid>
+      <Row className="well well-error">
+        <Col md={12} >
+          <h1 className="center">RPC connection Error (502)</h1>
```
Comment (rudolfix) [inline src/components/RPCProvider.js]: How do you know it is 502?

File: src/components/RPCProvider.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+          <p>Trying to connect to rpc node {config.rpcHost} received an invalid response.</p>
+          <a href="/" >Reload {props.counter}</a>
```
Comment (rudolfix) [inline src/components/RPCProvider.js]: here just block interface by modal and ask to reload the page, I think link is not necessary

File: src/components/ScanBoxDetails.js [hunk 9]
```diff
@@ -7,8 +7,10 @@ import { TimeDetails, RaisedAmount, TokenIssued, Investors } from './details';
 import { SingleBarChart } from './charts';
 import { default as config } from '../config.js';
 
-const ScanBoxDetails = ({ ...props }) => (props.currencyValue && <div className="scanbox-details">
+const ScanBoxDetails = ({ ...props }) => (<div className="scanbox-details">
+  {console.log(props)}
   <Row className="statistics">
+    {console.log('ScanBoxDetails component did mout')}
```
Comment (rudolfix) [inline src/components/ScanBoxDetails.js]: fix typo 'mout'

File: src/reducers/web3.js [hunk 10]
```diff
@@ -60,23 +56,27 @@ export const getLogs = address => async (dispatch, getState) => {
     dispatch(errorMessage());
     return;
   }
-  console.log('Start working on logs');
-  getICOLogs(web3, address, async (error, logs) => {
-    dispatch(hideLoader());
+  setCurrency('EUR', 'NOW', dispatch, () => {
```
Comment (rudolfix) [inline src/reducers/web3.js]: error callback is not handled, should be modal with error message

File: src/reducers/web3.js [hunk 11]
```diff
@@ -60,23 +56,27 @@ export const getLogs = address => async (dispatch, getState) => {
...
-    }
+      if (error || logs.length === 0) dispatch({ type: error });
+      else {
+        const currencyValue = getState().currency.value;
+        console.log('Fetched Currency is ', currencyValue);
+        const smartContractConstants = await getSmartContractConstants(web3, address);
+        const ico = config.ICOs[address];
+        ico.decimals = smartContractConstants.decimals;
+
+        if (currencyValue) {
```
Comment (rudolfix) [inline src/reducers/web3.js]: you do not need that. can you rename it to currencyRate?

File: src/actions/CurrencyAction.js (hunk 3 above)
Comment (MostafaBalata) [inline src/actions/CurrencyAction.js]: You already solved it in #17 `fix_decision_matrix` branch

**Pull request merged**